Design/Fit functions in python.md

```diff
@@ -0,0 +1,187 @@
+```
+  spectrum = LinearBackground() + Gaussian(PeakCentre=1) + Gaussian(PeakCentre=2)
+  peak1 = spectrum[1]
+  peak1['Sigma'] = 0.123
+  spectrum[2] = Lorentzian(PeakCentre=2)
+```
+
+`FitFunctionWrapper` will override `__iadd__(self, func)` and `__delitem__(self, i)` (`i` is an `int`, `func` is a `FitFunctionWrapper`) to implement adding and deleting members via `+=` and `del` operators:
+```
+  spectrum += Lorentzian(PeakCentre=3)
+  del spectrum[0]
+```
+
+Implement `__len__(self)` to return the number of member functions.
+```
+  n_peaks = len(spectrum)
+```
+
+### Setting ties
+
+Ties are set with the `tie` method. Both keyword arguments and dictionaries of name/value pairs can be used as arguments.
+```
+  func.tie(A0=2.0)
+  func.tie({'f1.A2': '2*f0.A1', 'f2.A2': '3*f0.A1 + 1'})
+```
+
+Tying to a constant can also be done with the `fix` method.
+```
+  func.fix('A0')             # Fix to the current value
+  func.fix(A0=2.0, A1=1.0)   # Fix to new values
+  func.fix({'f1.A2': 1.0, 'f2.A2': 2.0})
+```
+
+Ties are removed with the `removeTie` or `free` methods.
+```
+  func.free('f1.Sigma', 'f3.Sigma')
+  func.removeTie('f1.Sigma', 'f3.Sigma')
+```
+
+Composite functions should be able to set ties on all members with a single call:
+```
+  c.tieAll('Sigma')
+```
+The above call should tie all `Sigma` parameters in the member functions. It is equivalent to calling
+```
+  c.tie({'f1.Sigma': 'f0.Sigma', 'f2.Sigma': 'f0.Sigma', 'f3.Sigma': 'f0.Sigma', ...})
+```
+Similarly a call to `fixAll` fixes all parameters with the given name to their current value or to a new value if provided.
+```
+  spectrum.fixAll('FWHM')
+  spectrum.fixAll(FWHM=0.01)
+  spectrum.fixAll({'f1.FWHM': 0.01})
+```
+Calling `fixAll()`/`freeAll()` without arguments fixes/frees all parameters of a function.
+
+### Setting constraints
+
+Constraints are set with `constraints` method passing constraints as strings:
+```
+  g.constraints('-1 < PeakCentre < 1', 'Sigma > 0')
+```
+
+For composite functions constraints can be set both through the composite function and its members:
+```
+  c = LinearBackground() + Gaussian(PeakCentre=1) + Gaussian(PeakCentre=2)
+  c.constraints('f1.Sigma > 0')
+  c[2].constraints('0 < Sigma < 1')
+```
+
+Composite functions should be able to set constraints on all members with a single call:
+```
+  c.constrainAll('Height > 1', 'Sigma > 0')
+```
+This should constrain the named parameters in all members that have them. Members that don't have these parameters should be skipped.
+
+## Fitting
+
+Python function `Fit` should accept instances of `FitFunctionWrapper` as its `Function` argument. `Fit` does't modify the input function but the returned value has the output function with the fitted parameters. Instead of a tuple `Fit` will return an instance of class `FitOutput` that contains all outputs from the algorithm and gives access to them both via the index operator and python read-only properties:
+```
+  res = Fit(func, ws)
```

Comment:
We have recently started returning a [`namedtuple`](https://docs.python.org/2/library/collections.html#collections.namedtuple) from all Python algorithm functions (it will be mentioned at the next review). Will this already cover this?